### ISSUE TITLE
Mockery doesn't work as expected

### ISSUE THREAD

**Issue by noppanit** [request/request-promise#99]
Hi,

I'm trying to mock the whole `request-promise` as in the documentation. But my request still hits the real server. What should I do.

I'm using 

```
"mockery": "^1.4.1",
"request-promise": "^1.0.2",
```

Here's my code
```
    before(function before(done) {
      mockery.enable({
        warnOnReplace: false,
        warnOnUnregistered: false,
        useCleanCache: true
      });

      mockery.registerMock('request-promise', function () {
        var response = '{}';
        console.log('register'); // <-- This doesn't print out.
        return Bluebird.resolve(response.trim());
      });

      done();

    });

    after(function after(done) {
      mockery.disable();
      mockery.deregisterAll();
      done();
    });
```

This is how I call the test

```
      var request = {payload: {
        queryType: 'queryType'
      }};

      var expectQuery = {};
      analytics.query(request, function reply(response) {
        var data = response.data;
        var query = response.query;
        console.log(query);
        done();
      });
```

This is the code inside `analytics.query`
```
query: function druid(request, reply) {
    var query = request.payload;
    queries.query(query).then(function thenFunc(response) {
      return reply({
        data: response,
        query: query
      });
    })
    .catch(function catchFunc(err) {
      console.log('error'); // <-- It prints this
      return reply(Boom.notFound(err.error.error, {message: err.error.error}));
    });
  },
````

The request comes to `catch` function which I don't expect it as I thought `mockery` will catch any request and return `resolve`?

**Comment by analog-nico** [comment]
@florianschmidt1994 Maybe you can quickly spot the issue?

**Comment by analog-nico** [comment]
Sorry that this issue got stale. Is it still relevant or can I close it?

**Comment by vaelinn** [comment]
I am using mockery to mock request-promise as well. It is a bit dirtier than I would like in this instance(probably my own doing) but here is what I have for this situation that is working properly. 

`function setUpMocks(desiredResponse) {
  mockery.registerMock('request-promise', desiredResponse);
  mockery.registerAllowable('path-under-test', true);
  return require('path-under-test');
}

//Used in a test like so
    it.only('returns 200 success', () => {
      let response = {
        response: {
          body: {},
          statusCode: 200,
        },
      };

      const moduleUnderTest = setUpMocks(function request(){return Promise.resolve(response)});
      return moduleUnderTest.functionUnderTest(params...).then(result => {
        ...assertions
      });
    });
`
Hope that at least helps with your use case. Obviously I have generalized stuff in the code there and not included the describe and such from mocha. Certainly welcome any feedback on my usage but as is it is working for me.

**Comment by svox1** [comment]
> Sorry that this issue got stale. Is it still relevant or can I close it?
I habe the same problem:
```
mockery: 1.7.0
request-promise: 4.1.1
```

**Comment by analog-nico** [comment]
@svox1 Did you try the [example in the README](https://github.com/request/request-promise#mocking-request-promise)? Into which complications do you run?

**Comment by svox1** [comment]
@analog-nico Yes. Direct in one test it works as expected but on a complex structure were the calls cam later (starting a express server) it didnt work.

Now I used https://github.com/node-nock/nock and it works like a charm :)

**Comment by analog-nico** [comment]
Thanks, good to know. I'll leave this issue open until I added better examples to the docs.

If you have a code snippet for how to use Nock it would be a great starting point for me. Thanks!

**Comment by noppanit** [comment]
I've tried `nock` and it works perfectly. This is what I do in my test

```
afterEach(() => {
  nock.cleanAll();
});

nock('/api')
       .get(`/user`)
       .reply(200, {results: []});
```

**Comment by noppanit** [comment]
I've switched to `nock` and it works for me. Please feel free to close the ticket.

**Comment by analog-nico** [comment]
Thanks for your code snippet @noppanit ! I'll keep this issue open though as a reminder once I update the docs.

**Comment by petrgazarov** [comment]
Hey @analog-nico, `nock` works perfectly for me as well. Do you still want to update the docs with `nock` example? Happy to send PR.

**Comment by analog-nico** [comment]
@petrgazarov Yes, that would be awesome. Thanks!

**Comment by JimLynchCodes** [comment]
Hello @petrgazarov @analog-nico @noppanit @svox1 @vaelinn,

Sorry to dig up this old thread, but I just had to jump in because I am having issues with mocking the `request-promise` module and do not believe that `nock` is the solution here. First of all (nock)[https://github.com/nock/nock] is an "HTTP Mocking for Node.js" which is not at all the same as `mockery` which is for mocking arbitrary libraries that are require'd in files outside of your test. I would prefer to use mockery to simply mock the required module, but my code similar to @vaelinn's is not working (it still makes a request use the real library).

Here is a live demo: https://repl.it/@JimLynch/Request-Promise-Mockery-Example

And the source code here: https://github.com/JimLynchCodes/Request-Promise-Mockery-Example

It would be great if someone could help me out with how to mock this required `request-promise` module. 🙏

Thanks!

**Comment by vaelinn123** [comment]
@JimLynchCodes It looks like you may need to enable mockery in your test. Try adding:

```
mockery.enable({
    warnOnReplace: false,
    warnOnUnregistered: false
});
```
on line 6 [here](https://github.com/JimLynchCodes/Request-Promise-Mockery-Example/blob/master/callForStuff.test.js#L6).

If that does not work, I can try to pull down your code and see if i can get it working locally. I won't be able to do that until tomorrow morning(Currently 6:30 PM where I am for reference)

**Comment by JimLynchCodes** [comment]
THANK YOU @vaelinn ! 🙏

Yes, it turns out I just needed to "enable" mockery in my tests. doh! 😁

Also, I noticed that this line that calls `registerAllowable` doesn't seem to do affect the code execution at all...  
```
mockery.registerAllowable('./callForStuff', true);
``` 

Thanks all!

**Comment by vaelinn123** [comment]
Awesome, glad to hear it worked out. I cannot remember what I was using registerAllowable for, was probably something specific to context outside of my example.

**Comment by vaelinn123** [comment]
I don't have access to that vaelinn account anymore, so here is an updated version of that snippet with the included mockery enable and removal of the registerAllowable bit:

```
function setUpMocks(desiredResponse) {
  mockery.enable({
      warnOnReplace: false,
      warnOnUnregistered: false
  });
  mockery.registerMock('request-promise', desiredResponse);
  return require('path-under-test');
}

//Used in a test like so
    it.only('returns 200 success', () => {
      let response = {
        response: {
          body: {},
          statusCode: 200,
        },
      };

      const moduleUnderTest = setUpMocks(function request(){return Promise.resolve(response)});
      return moduleUnderTest.functionUnderTest(params...).then(result => {
        ...assertions
      });
    });
```
